File: nim/grade-school/grade_school.nim
import algorithm, sequtils

type
  Student = tuple[name: string, grade: int]

  School* = object
    students*: seq[Student]

proc roster*(school: School): seq[string] =
  school.students.sortedByIt((it.grade, it.name)).mapIt(it.name)

proc grade*(school: School, grade: int): seq[string] =
  school.students.filterIt(it.grade == grade).mapIt(it.name).sorted

File: clojure/grade-school/src/grade_school.clj
(ns grade-school)

(defn grade [school grade]
  (get school grade []))

(defn add [school name grade]
  (merge-with concat school {grade [name]}))

(defn sorted [school]
  (into (sorted-map) (for [[grade names] school] [grade (sort names)])))
